function/expression.go

```diff
@@ -23,6 +24,35 @@ type EvaluationContext struct {
 	FetchLimit   FetchCounter     // A limit on the number of fetches which may be performed
 	Cancellable  api.Cancellable
 	Profiler     *inspect.Profiler
+	Registry     Registry
+}
+
+type Registry interface {
+	GetFunction(string) (MetricFunction, bool)
+}
+
+// MetricFunction defines a common logic to dispatch a function in MQE.
+type MetricFunction struct {
+	Name        string
+	MinArgument int
+	MaxArgument int
+	Groups      bool // Whether the function allows a 'group by' clause.
+	Compute     func(EvaluationContext, []Expression, []string) (Value, error)
+}
+
+// Evaluate the given metric function.
+func (f MetricFunction) Evaluate(context EvaluationContext,
+	arguments []Expression, groupBy []string) (Value, error) {
+	// preprocessing
+	length := len(arguments)
+	if length < f.MinArgument || (f.MaxArgument != -1 && f.MaxArgument < length) {
+		return nil, ArgumentLengthError{f.Name, f.MinArgument, f.MaxArgument, length}
```

Comment:
I don't think we have argument names currently.

This PR is mostly move, so I'll add it later.